### PR TITLE
[Failing Codecov only] Platforms: Config Upgrader

### PR DESCRIPTION
**Pull request by wxtim**
Function to upgrade the suite configuration objects to use platform rather than various settings in `[runtime][task]`.

- [x] written a function to do this.
- [x] written tests for this function.

### CODE REVIEW

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
> written functional tests for this function.

Doesn't seem to be included in the PR?

**Comment by wxtim** [comment]
> > written functional tests for this function.
> 
> Doesn't seem to be included in the PR?

That's why it's a WIP PR.

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  8
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

File: cylc/flow/cfgspec/suite.py [hunk 1]
```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
...
+    # upgrader
+    # cfg = host_to_platform(cfg)
+
+
+def host_to_platform_upgrader(cfg):
+    """Upgrade a config with host settings to a config with platform settings
+    if it is appropriate to do so.
+
+                       +-------------------------------+
+                       | Is platforms set in this      |
```
Comment (oliver-sanders) [inline cylc/flow/cfgspec/suite.py]:
```suggestion
                       | Is platform set in this       |
```

File: cylc/flow/cfgspec/suite.py [hunk 2]
```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
...
+        #     breakpoint(header=f"task_name = {task_name}")
+
+        if (
+            'platform' in task_spec and 'job' in task_spec or
+            'platform' in task_spec and 'remote' in task_spec
+        ):
+            if (
+                'platform' in task_spec and
+                any(forbidden_with_platform == task_spec['job']) or
+                any(forbidden_with_platform == task_spec['remote'])
```
Comment (oliver-sanders) [inline cylc/flow/cfgspec/suite.py]: These `any` comparisons don't look right, should they be set comparisons:

```
forbidden_with_platform = { 'host', ... }
# ...
forbidden_with_platform & {*task_spec['job'], *task_spec['remote']}
```

File: cylc/flow/cfgspec/suite.py [hunk 3]
```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
...
+            # Attempt to use the reverse lookup
+            try:
+                platform = reverse_lookup(
+                    glbl_cfg(cached=False).get(['job platforms']),
+                    task_spec_job,
+                    task_spec_remote
+                )
+            except PlatformLookupError as exc:
+                LOG.error(f'Unable to determine platform for {task_name}')
+                LOG.debug(f'Exception was {exc}')
```
Comment (oliver-sanders) [inline cylc/flow/cfgspec/suite.py]: If we can't determine a platform we can't run the suite so I guess just let this error float to the top or re-raise it as a `ConfigError` with more helpful text.

Comment (wxtim) [reply cylc/flow/cfgspec/suite.py]: Do you mean use Python exceptions rather than the Cylc Logging?

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  14
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

File: cylc/flow/cfgspec/suite.py (hunk 2 above)
Comment (wxtim) [reply cylc/flow/cfgspec/suite.py]: Well for one thing this comment made me realise that I wasn't testing this logic!

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

File: cylc/flow/cfgspec/suite.py (hunk 3 above)
Comment (oliver-sanders) [reply cylc/flow/cfgspec/suite.py]: Yes, if we don't fail loudly here it will become a runtime error later on.

Comment (wxtim) [reply cylc/flow/cfgspec/suite.py]: Soundeth Good. Will do.

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by oliver-sanders** [comment]
I think there may have been an issue with codecov upload, I ran coverage locally:

```console
$ pytest --cov cylc/flow
$ codecov html -d output
$ <web-browser> output/index.html
```

There are a couple of branches which are untested:

* Return config unchanged
* if 'job' **not** in task_spec
* Platform {platform} auto selected from

The latter branch really needs testing.

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  15
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

File: cylc/flow/cfgspec/suite.py [hunk 4]
```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
...
+            if (
+                'host' in task_spec_remote and
+                REC_COMMAND.match(task_spec['remote']['host'])
+            ):
+                LOG.info(
+                    f"Unable to upgrade task '{task_name}' to platform at "
+                    f"validation because the host setting contains a "
+                    f"function. Cylc will attempt to upgrade this task on"
+                    f" job submission."
+                )
```
Comment (oliver-sanders) [inline cylc/flow/cfgspec/suite.py]: For a suite with ~10'000 tasks this logging may be a bit overwhelming. We can address logging closer to release time...

File: cylc/flow/cfgspec/suite.py [hunk 5]
```diff
@@ -335,9 +344,135 @@ def upg(cfg, descr):
...
+            'platform' in task_spec and 'remote' in task_spec
+        ):
+            if (
+                'platform' in task_spec and
+                forbidden_with_platform & {
+                    *task_spec['job'], *task_spec['remote']
+                }
+            ):
+                # Fail Loudly and Horribly
+                LOG.error(
```
Comment (oliver-sanders) [inline cylc/flow/cfgspec/suite.py]: This also needs to be changed to an exception if we want it to actually fail.

File: cylc/flow/cfgspec/suite.py (hunk 4 above)
Comment (wxtim) [reply cylc/flow/cfgspec/suite.py]: Exception instead?

Comment (wxtim) [reply cylc/flow/cfgspec/suite.py]: Exception instead?

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  15
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

File: cylc/flow/cfgspec/suite.py (hunk 4 above)
Comment (wxtim) [reply cylc/flow/cfgspec/suite.py]: I've made this a debug statement and streamlined the text - it's now only 2 lines long!

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  15
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by hjoliver** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- cylc/flow/cfgspec/suite.py  15
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Cylc/cylc-flow/pullRequest?prid=4673368&bid=15638667)

**Comment by wxtim** [comment]
> Great, could do with a rebase-squash before merge, otherwise second reviewer can merge-squash.

The history is a little complex and I'm not sure how possible this will be - I think merge squash may work better.

**Comment by wxtim** [comment]
can I poke @hjoliver to have a look, otherwise will go through this with @datamel on Monday and seek her approval then merge.

**Comment by oliver-sanders** [comment]
You can try but he's been off grid since Wednesday, will have to wait.

**Pull request merged**